api/inc/pool_queue_exports.h

```diff
@@ -0,0 +1,201 @@
+     * from the pool. If non-zero, when no slots is available in the pool,
+     * callers will be blocked up to their timeout amount of time before giving
+     * up. */
+    int blocking;
+
+    /* The number of items currently allocated from the pool. For testing and
+     * debug purposes only. */
+    uvisor_pool_slot_t num_allocated;
+
+    /* The first free slot. */
+    uvisor_pool_slot_t first_free;
+
+    /* The semaphore is used to block allocations when the pool is full. */
+    UvisorSemaphore semaphore;
+
+    /* The spinlock serializes updates to the management array. */
+    UvisorSpinlock spinlock;
+
+    /* This must be at the end so we can allocate memory for pools by
+     * allocating enough room for the size of the pool appended by an array of
+     * entries. */
+    uvisor_pool_queue_entry_t management_array[];
+} uvisor_pool_t;
+
+typedef struct uvisor_pool_queue {
+    /* The first allocated slot */
+    uvisor_pool_slot_t head;
+
+    /* The last allocated slot */
+    uvisor_pool_slot_t tail;
+
+    uvisor_pool_t * pool;
+} uvisor_pool_queue_t;
+
+/* Intialize a pool.
+ * Return 0 on success, non-zero otherwise. */
+UVISOR_EXTERN int uvisor_pool_init(uvisor_pool_t * pool, void * array, size_t stride, size_t num, int blocking);
+
+/* Initialize a pool queue.
+ * Return 0 on success, non-zero otherwise. */
+UVISOR_EXTERN int uvisor_pool_queue_init(uvisor_pool_queue_t * pool_queue, uvisor_pool_t * pool, void * array, size_t stride, size_t num, int blocking);
+
+/* Allocate a slot from the pool. If the pool has no more slots available,
+ * block up to the specified length of time in milliseconds. No blocking will
+ * occur if the timeout is zero or the pool was initialized as non-blocking.
+ * This doesn't put anything in the slot for you. It's up to you to do that.
+ * Return the index of the allocated slot, or UVISOR_POOL_SLOT_INVALID if
+ * timed out waiting for an available slot. This function will spin until the
+ * spin lock serializing access to the pool can be taken. */
+UVISOR_EXTERN uvisor_pool_slot_t uvisor_pool_allocate(uvisor_pool_t * pool, uint32_t timeout_ms);
+/* Attempt to allocate a slot. This function will fail if the spin lock
+ * serializing access to the pool can not be taken. */
+UVISOR_EXTERN uvisor_pool_slot_t uvisor_pool_try_allocate(uvisor_pool_t * pool);
+
+/* Enqueue the specified slot into the queue. */
+UVISOR_EXTERN void uvisor_pool_queue_enqueue(uvisor_pool_queue_t * pool_queue, uvisor_pool_slot_t slot);
+UVISOR_EXTERN int uvisor_pool_queue_try_enqueue(uvisor_pool_queue_t * pool_queue, uvisor_pool_slot_t slot);
+
+/* Free the specified slot back into the pool. Invalid slots are ignored.
+ * Return the slot that was freed, or UVISOR_POOL_SLOT_IS_FREE if the slot was
+ * already freed, or UVISOR_POOL_SLOT_INVALID if the slot being requested to
+ * free is outside the range of the queue. */
+UVISOR_EXTERN uvisor_pool_slot_t uvisor_pool_free(uvisor_pool_t * pool, uvisor_pool_slot_t slot);
```

Comment:
There are multiple queues per box context: incoming RPC message queue and outgoing result queue.